File: nekoshlink/shlink-security/src/main/java/org/nekosoft/shlink/sec/web/ShlinkAdminApiFormSecurityConfiguration.java
package org.nekosoft.shlink.sec.web;

import org.springframework.boot.autoconfigure.condition.ConditionalOnWebApplication;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.core.annotation.Order;
import org.springframework.security.config.annotation.web.builders.HttpSecurity;
import org.springframework.security.web.SecurityFilterChain;

//@Configuration
@ConditionalOnWebApplication
public class ShlinkAdminApiFormSecurityConfiguration {

    @Bean
    @Order(99)
    public SecurityFilterChain filterChainBasicAdminApiLogin(HttpSecurity http) throws Exception {
        return http
                .requestMatchers().mvcMatchers(
                        "/login*",
                        "/logout*"
                )
                .and()
                .authorizeRequests(authorize ->
                        authorize.anyRequest().authenticated()
                )
                .formLogin()
                .and()
                .logout()
                .and().build();
    }

    @Bean
    @Order(120)
    public SecurityFilterChain filterChainBasicAdminApi(HttpSecurity http) throws Exception {
        return http
                .cors()
                .and()
                .authorizeRequests(authorize ->
                        authorize.anyRequest().authenticated()
                )
                .formLogin()
                .and().build();
    }

}

File: nekoshlink/shlink-security/src/main/kotlin/org/nekosoft/shlink/sec/web/ShlinkAdminApiFormSecurityConfiguration.kt
package org.nekosoft.shlink.sec.web

import org.springframework.boot.autoconfigure.condition.ConditionalOnWebApplication
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.core.annotation.Order
import org.springframework.http.HttpMethod
import org.springframework.security.config.annotation.web.builders.HttpSecurity
import org.springframework.security.config.web.servlet.invoke
import org.springframework.security.web.SecurityFilterChain


@Configuration
@ConditionalOnWebApplication
class ShlinkAdminApiFormSecurityConfiguration {

    @Bean
    @Order(99)
    fun filterChainBasicAdminApiLogin(http: HttpSecurity): SecurityFilterChain {
        http {
            securityMatcher(
                "/login*",
                "/logout*",
            )
            formLogin { }
            logout {  }
        }
        return http.build()
    }

    @Bean
    @Order(120)
    fun filterChainBasicAdminApi(http: HttpSecurity): SecurityFilterChain {
        http {
            cors {  }
            authorizeRequests {
                authorize(HttpMethod.GET, "/api/v1/shorturls/**", hasAnyRole("Admin", "Editor", "Viewer"))
                authorize(HttpMethod.POST, "/api/v1/shorturls/**", hasAnyRole("Admin", "Editor"))
                authorize(HttpMethod.PUT, "/api/v1/shorturls/**", hasAnyRole("Admin", "Editor"))
                authorize(HttpMethod.DELETE, "/api/v1/shorturls/**", hasRole("Admin"))

                authorize(HttpMethod.GET, "/api/v1/domains/**", hasRole("Admin"))
                authorize(HttpMethod.POST, "/api/v1/domains/**", hasRole("Admin"))
                authorize(HttpMethod.PUT, "/api/v1/domains/**", hasRole("Admin"))
                authorize(HttpMethod.PATCH, "/api/v1/domains/**", hasRole("Admin"))
                authorize(HttpMethod.DELETE, "/api/v1/domains/**", hasRole("Admin"))

                authorize(HttpMethod.GET, "/api/v1/tags/**", hasAnyRole("Admin", "Editor", "Viewer"))
                authorize(HttpMethod.POST, "/api/v1/tags/**", hasAnyRole("Admin", "Editor"))
                authorize(HttpMethod.PUT, "/api/v1/tags/**", hasAnyRole("Admin", "Editor"))
                authorize(HttpMethod.PATCH, "/api/v1/tags/**", hasAnyRole("Admin", "Editor"))
                authorize(HttpMethod.DELETE, "/api/v1/tags/**", hasRole("Admin"))

                authorize(HttpMethod.GET, "/api/v1/visits/**", hasRole("Admin"))

                authorize(anyRequest, authenticated)
            }
            formLogin {  }
        }
        return http.build()
    }

}
